### DISCUSSION TITLE
CSP Error with script history.replaceState() During Flow Transitions (Keycloak 23.0.4) when I remove unsafe-inline

### DISCUSSION THREAD

**Discussion by Bart094** [keycloak/keycloak#30074] [Q&A]
Hello Keycloak community,

I'm facing a Content Security Policy (CSP) error in Keycloak 23.0.4 when attempting to remove unsafe-inline. The error is triggered by a script that Keycloak injects, which utilizes history.replaceState(). This issue specifically arises during transitions between different authentication flows (e.g., from the login page to the username recovery page).

The problematic script appears to be dynamically generated by Keycloak's backend, making it difficult to apply standard CSP workarounds like nonces or hashes.

Expected Behavior:
Keycloak should function correctly with a strict CSP that prohibits unsafe-inline, without generating errors during flow transitions.

Any insights, suggestions, or known workarounds?
Thank you!

**Comment by JanBuecher** [comment]
Still a problem in version 26.0.5

**Comment by LaughDonor** [comment]
Same with 26.1.3. We are using the default CSP within Realm Settings -> Security Defenses: `frame-src 'self'; frame-ancestors 'self'; object-src 'none';`
![image](https://github.com/user-attachments/assets/2a3e7c02-2a40-4b8a-a9ca-f20379026358)

This happens during authentication flow transitions between subdomains. I want the browser to be able to process this code block, but haven't been able to figure out a workaround through updating Keycloak's CSP using `script-src` or `script-src-elem` props. If it worked, I would assume that Keycloak could inject a nonce prop to match this injection.

https://github.com/keycloak/keycloak/blob/c11d93d93634b19102387bb89858c3043ccb40b9/services/src/main/java/org/keycloak/services/util/BrowserHistoryHelper.java#L111-L135

**Comment by weeks6** [comment]
Did anyone manage to resolve this?
